pygsheets/datafilter.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-.
+
+"""
+pygsheets.datafilter
+~~~~~~~~~~~~~~~~~~~~~
+
+This module represents a datafilter that describes what data should be selected
+or returned from a request.
+
+"""
+
+import logging
+import warnings
+
+
+class DataFilter(object):
+    """Base class for DataFilter types"""
+    def serialize(self):
+        raise NotImplemented()
+
+
+class A1RangeDataFilter(DataFilter):
+    """Class for filtering data based on an A1 range
+
+    :param a1_range:  The cell range to filter on
+    """
+    def __init__(self, a1_range):
+        self.a1_range = a1_range
+
+    def serialize(self):
+        return {"a1Range": self.a1_range}
+
+
+class GridRangeDataFilter(DataFilter):
+    """Class for filtering data based on start and end indices
+
+    This method uses zero-based indices. Start indices must be less
+    than their corrisponding end index.
+
+    :param sheet_id:            Worksheet id to filter on
+    :param start_row_index:     Start row of the range to filter on
+    :param end_row_index:       End row of the range to filter on
+    :param start_column_index:  Start column of the range to filter on
+    :param end_column_index:    End end of the range to filter on
+    """
+    def __init__(self, sheet_id, start_row_index, end_row_index, start_column_index, end_column_index):
+        #TODO: make sure start and end indexes are > 0 and start < end
+        self.sheet_id = sheet_id
+        self.start_row_index = start_row_index
+        self.end_row_index = end_row_index
+        self.start_column_index = start_column_index
+        self.end_column_index = end_column_index
+
+    def serialize(self):
+        return {
+            "gridRange": {
+                "sheetId": self.sheet_id,
+                "startRowIndex": self.start_row_index,
+                "endRowIndex": self.end_row_index,
+                "startColumnIndex": self.start_column_index,
+                "endColumnIndex": self.end_column_index
+            }
+        }
+
+
+class DeveloperMetadataLookupDataFilter(DataFilter):
```

Comment:
Both of those other classes are intended to be used with the `spreadsheets.getByDataFilter` rather than to represent an a1 range or gridrange.

https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/getByDataFilter

If that's not useful, I'll just pull the DeveloperMetadataLookupDataFilter into developer_metadata.py like you suggest.

Reply:
form api docs, what I understand is the `a1Range` and `gridRange` params lets the user specify a range of cells in the sheet. just two different formats. But in our case we already have a wrapper to represent the range. The GridRange object is flexible enough to let the user specify the range as A1, start and end cells, cell labels, gridRange format etc.  

this will remove code duplication since the gridRange already does validation checks

so while implementing filtering, we can just take a single param which is GridRange to represent range of cells. where they could provide the address any format they intend to. and we can use [this](https://github.com/nithinmurali/pygsheets/blob/staging/pygsheets/address.py#L348).

Reply:
Sounds good. I'll implement the changes you requested.